roachtest: tpchvec/disk failed

[(roachtest).tpchvec/disk failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1834610&tab=buildLog) on [master@0e9f07ae35ad39e8789bc51c49066c123968a666](https://github.com/cockroachdb/cockroach/commits/0e9f07ae35ad39e8789bc51c49066c123968a666):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200328-1834610/tpchvec/disk/run_1
	tpchvec.go:579,tpchvec.go:669,test_runner.go:753: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2031
		    main.registerTPCHVec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:574
		    main.registerTPCHVec.func3
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:669
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1834610-1585381212-20-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=1 --queries=2 --vectorize=on {pgurl:1-3}:
		  - exit status 30

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec/disk](https://teamcity.cockroachdb.com/viewLog.html?buildId=1834610&tab=artifacts#/tpchvec/disk)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec%2Fdisk.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchvec/disk failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1853455&tab=buildLog) on [master@b1a0b989bbfef500075a485edc762fe42ca7b32a](https://github.com/cockroachdb/cockroach/commits/b1a0b989bbfef500075a485edc762fe42ca7b32a):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200405-1853455/tpchvec/disk/run_1
	tpchvec.go:578,tpchvec.go:670,test_runner.go:753: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2031
		    main.registerTPCHVec.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:573
		    main.registerTPCHVec.func4
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:670
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1853455-1586071414-19-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=1 --queries=2 --vectorize=on {pgurl:1-3}:
		  - exit status 30

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec/disk](https://teamcity.cockroachdb.com/viewLog.html?buildId=1853455&tab=artifacts#/tpchvec/disk)
Related:
- #46810 roachtest: tpchvec/disk failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec%2Fdisk.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchvec/disk failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1854541&tab=buildLog) on [master@beac4a53e0e2e2236eb5957f67abc1bf476ad1b6](https://github.com/cockroachdb/cockroach/commits/beac4a53e0e2e2236eb5957f67abc1bf476ad1b6):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200406-1854541/tpchvec/disk/run_1
	tpchvec.go:578,tpchvec.go:670,test_runner.go:753: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2031
		    main.registerTPCHVec.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:573
		    main.registerTPCHVec.func4
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:670
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1854541-1586157738-19-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=1 --queries=2 --vectorize=on {pgurl:1-3}:
		  - exit status 30

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec/disk](https://teamcity.cockroachdb.com/viewLog.html?buildId=1854541&tab=artifacts#/tpchvec/disk)
Related:
- #46810 roachtest: tpchvec/disk failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec%2Fdisk.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchvec/disk failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1857109&tab=buildLog) on [master@2032dafccfa311c7538960e974953cb9dc1d4e50](https://github.com/cockroachdb/cockroach/commits/2032dafccfa311c7538960e974953cb9dc1d4e50):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200407-1857109/tpchvec/disk/run_1
	tpchvec.go:578,tpchvec.go:670,test_runner.go:753: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2031
		    main.registerTPCHVec.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:573
		    main.registerTPCHVec.func4
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:670
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1857109-1586244496-19-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=1 --queries=2 --vectorize=on {pgurl:1-3}:
		  - exit status 30

	cluster.go:1410,context.go:135,cluster.go:1399,test_runner.go:825: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1857109-1586244496-19-n3cpu4 --oneshot --ignore-empty-nodes: exit status 1 1: dead
		3: dead
		2: dead
		Error: UNCLASSIFIED_PROBLEM:
		  - 1: dead, 3: dead, 2: dead
		    main.glob..func13
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1129
		    main.wrap.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:272
		    github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).execute
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:766
		    github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).ExecuteC
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:852
		    github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).Execute
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:800
		    main.main
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1793
		    runtime.main
		    	/usr/local/go/src/runtime/proc.go:203
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec/disk](https://teamcity.cockroachdb.com/viewLog.html?buildId=1857109&tab=artifacts#/tpchvec/disk)
Related:
- #46810 roachtest: tpchvec/disk failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec%2Fdisk.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchvec/disk failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1862625&tab=buildLog) on [master@0222b515560bb02e6adf59a09f6c067923bca28c](https://github.com/cockroachdb/cockroach/commits/0222b515560bb02e6adf59a09f6c067923bca28c):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200409-1862625/tpchvec/disk/run_1
	tpchvec.go:578,tpchvec.go:670,test_runner.go:753: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2041
		    main.registerTPCHVec.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:573
		    main.registerTPCHVec.func4
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:670
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1862625-1586418509-19-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=1 --queries=2 --vectorize=on {pgurl:1-3}:
		  - exit status 30

	cluster.go:1420,context.go:135,cluster.go:1409,test_runner.go:825: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1862625-1586418509-19-n3cpu4 --oneshot --ignore-empty-nodes: exit status 1 1: 15836
		3: dead
		2: 14195
		Error: UNCLASSIFIED_PROBLEM:
		  - 3: dead
		    main.glob..func13
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1129
		    main.wrap.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:272
		    github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).execute
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:766
		    github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).ExecuteC
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:852
		    github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra.(*Command).Execute
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:800
		    main.main
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1793
		    runtime.main
		    	/usr/local/go/src/runtime/proc.go:203
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec/disk](https://teamcity.cockroachdb.com/viewLog.html?buildId=1862625&tab=artifacts#/tpchvec/disk)
Related:
- #46810 roachtest: tpchvec/disk failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec%2Fdisk.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).tpchvec/disk failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1866229&tab=buildLog) on [master@2b2eac926b5a51f78d64f6d7d3e86c80c13931a4](https://github.com/cockroachdb/cockroach/commits/2b2eac926b5a51f78d64f6d7d3e86c80c13931a4):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200410-1866229/tpchvec/disk/run_1
	tpchvec.go:578,tpchvec.go:670,test_runner.go:753: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2041
		    main.registerTPCHVec.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:573
		    main.registerTPCHVec.func4
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:670
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1866229-1586504087-21-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=1 --queries=2 --vectorize=on {pgurl:1-3}:
		  - exit status 30

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec/disk](https://teamcity.cockroachdb.com/viewLog.html?buildId=1866229&tab=artifacts#/tpchvec/disk)
Related:
- #46810 roachtest: tpchvec/disk failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec%2Fdisk.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/cmd/roachtest/tpchvec.go
+++ b/pkg/cmd/roachtest/tpchvec.go
@@ -540,8 +540,8 @@ RESTORE tpch.* FROM 'gs://cockroach-fixtures/workload/tpch/scalefactor=1/backup'
 		if runConfig.stressDiskSpilling {
 			// In order to stress the disk spilling of the vectorized
 			// engine, we will set workmem limit to a random value in range
-			// [100KiB, 1000KiB).
-			workmemInKiB := 100 + rng.Intn(900)
+			// [16KiB, 256KiB).
+			workmemInKiB := 16 + rng.Intn(240)
 			workmem := fmt.Sprintf("%dKiB", workmemInKiB)
 			t.Status(fmt.Sprintf("setting workmem='%s'", workmem))
 			if _, err := conn.Exec(fmt.Sprintf("SET CLUSTER SETTING sql.distsql.temp_storage.workmem='%s'", workmem)); err != nil {

--- a/pkg/sql/colexec/external_hash_joiner.go
+++ b/pkg/sql/colexec/external_hash_joiner.go
@@ -116,6 +116,9 @@ const (
 	//   has returned a zero batch, and thus the FD has been closed.
 	sortMergeNonSortMinFDsOpen = 2
 	externalHJMinPartitions    = sortMergeNonSortMinFDsOpen + (externalSorterMinPartitions * 2)
+	// externalHJMinimalMaxRightPartitionSize determines the minimum value for
+	// maxRightPartitionSizeToJoin variable of the external hash joiner.
+	externalHJMinimalMaxRightPartitionSize = 64 << 10 /* 64 KiB */
 )
 
 // externalHashJoiner is an operator that performs Grace hash join algorithm
@@ -389,6 +392,9 @@ func newExternalHashJoiner(
 	// output batch will be used, but that shouldn't matter in the grand scheme
 	// of things.
 	ehj.memState.maxRightPartitionSizeToJoin = memoryLimit - int64(diskQueuesMemUsed)
+	if ehj.memState.maxRightPartitionSizeToJoin < externalHJMinimalMaxRightPartitionSize {
+		ehj.memState.maxRightPartitionSizeToJoin = externalHJMinimalMaxRightPartitionSize
+	}
 	ehj.scratch.leftBatch = unlimitedAllocator.NewMemBatch(spec.left.sourceTypes)
 	ehj.recursiveScratch.leftBatch = unlimitedAllocator.NewMemBatchNoCols(spec.left.sourceTypes, 0 /* size */)
 	sameSourcesSchema := len(spec.left.sourceTypes) == len(spec.right.sourceTypes)

